src/main/java/uk/gov/ida/matchingserviceadapter/services/VerifyAssertionService.java

```diff
@@ -0,0 +1,104 @@
+package uk.gov.ida.matchingserviceadapter.services;
+
+import org.opensaml.saml.saml2.core.Assertion;
+import org.opensaml.saml.saml2.core.AuthnStatement;
+import org.opensaml.saml.saml2.metadata.IDPSSODescriptor;
+import uk.gov.ida.matchingserviceadapter.domain.AssertionData;
+import uk.gov.ida.matchingserviceadapter.exceptions.SamlResponseValidationException;
+import uk.gov.ida.matchingserviceadapter.validators.ConditionsValidator;
+import uk.gov.ida.matchingserviceadapter.validators.InstantValidator;
+import uk.gov.ida.matchingserviceadapter.validators.SubjectValidator;
+import uk.gov.ida.saml.core.transformers.AuthnContextFactory;
+import uk.gov.ida.saml.core.transformers.VerifyMatchingDatasetUnmarshaller;
+import uk.gov.ida.saml.core.transformers.inbound.Cycle3DatasetFactory;
+import uk.gov.ida.saml.security.SamlAssertionsSignatureValidator;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.emptyList;
+import static uk.gov.ida.saml.core.validation.errors.GenericHubProfileValidationSpecification.MISMATCHED_ISSUERS;
+import static uk.gov.ida.saml.core.validation.errors.GenericHubProfileValidationSpecification.MISMATCHED_PIDS;
+
+public class VerifyAssertionService extends AssertionService {
+
+    private String hubEntityId;
+    private static final String AUTHN_ASSERTION = "authn-assertion";
+    private static final String MDS_ASSERTION = "mds-assertion";
+    private static final String CYCLE_3_ASSERTION = "cycle-3-assertion";
+    private final VerifyMatchingDatasetUnmarshaller matchingDatasetUnmarshaller;
+    private AuthnContextFactory authnContextFactory = new AuthnContextFactory();
+
+    public VerifyAssertionService(InstantValidator instantValidator,
+                                  SubjectValidator subjectValidator,
+                                  ConditionsValidator conditionsValidator,
+                                  SamlAssertionsSignatureValidator hubSignatureValidator,
+                                  Cycle3DatasetFactory cycle3DatasetFactory,
+                                  String hubEntityId,
+                                  VerifyMatchingDatasetUnmarshaller matchingDatasetUnmarshaller) {
+        super(instantValidator, subjectValidator, conditionsValidator, hubSignatureValidator, cycle3DatasetFactory);
+        this.hubEntityId = hubEntityId;
+        this.matchingDatasetUnmarshaller = matchingDatasetUnmarshaller;
+    }
+
+    @Override
+    public void validate(String requestId, List<Assertion> assertions) {
+        Map<String, List<Assertion>> assertionMap = assertions.stream()
+                .collect(Collectors.groupingBy(this::classifyAssertion));
+
+        List<Assertion> authnAssertions = assertionMap.get(AUTHN_ASSERTION);
+        if (authnAssertions.isEmpty() || authnAssertions.size() != 1) {
```

Comment:
I think `.get()` is going to return null, not the empty list if there aren't any `AUTHN_ASSERTION`s, so there's a potential NPE here. Also the `.isEmpty()` check is redundant given that you're already checking that `size() != 1`.

Reply:
Yup, found this one in testing. Updated in the rewrite branch. Made the same change in the mdsAssertions check 5 lines below too.